src/vec.rs

```diff
@@ -1,43 +1,53 @@
+//! A fixed capacity [`Vec`](https://doc.rust-lang.org/std/vec/struct.Vec.html).
+
 use core::{
+    borrow::{Borrow, BorrowMut},
     cmp::Ordering,
-    fmt, hash, mem,
-    mem::{ManuallyDrop, MaybeUninit},
+    fmt, hash,
+    mem::{self, ManuallyDrop, MaybeUninit},
     ops, ptr, slice,
 };
 
-pub trait VecBuffer {
-    type T;
-
-    fn as_vecview(vec: &VecInner<Self>) -> &VecView<Self::T>;
-    fn as_mut_vecview(vec: &mut VecInner<Self>) -> &mut VecView<Self::T>;
+trait SealedStorage {
+    type Buffer<T>: ?Sized + Borrow<[MaybeUninit<T>]> + BorrowMut<[MaybeUninit<T>]>;
 }
 
-impl<T, const N: usize> VecBuffer for [MaybeUninit<T>; N] {
-    type T = T;
-
-    fn as_vecview(vec: &VecInner<Self>) -> &VecView<Self::T> {
-        vec
-    }
-    fn as_mut_vecview(vec: &mut VecInner<Self>) -> &mut VecView<Self::T> {
-        vec
-    }
+/// Trait defining how the data for a Vec is stored.
+///
+/// There's two implementations available:
+///
+/// - [`VecStorage`]: the storage for a [`Vec`]. Stores the data in an array `[MaybeUninit<T>; N]` whose size is known at compile time.
+/// - [`VecViewStorage`]: the storage for a [`VecView`]. Stores the data in an unsized `[MaybeUninit<T>]`.
+///
+/// This allows [`VecInner`] (the base struct for vectors) to be generic over either sized
+/// or unsized storage.
+///
+/// This trait is sealed, so you cannot implement it for your own types. You can only use
+/// the implementations provided by this crate.
+#[allow(private_bounds)]
+pub trait Storage: SealedStorage {}
+
+/// Implementation of [`Storage`] for [`Vec`].
+pub enum VecStorage<const N: usize> {}
+impl<const N: usize> Storage for VecStorage<N> {}
+impl<const N: usize> SealedStorage for VecStorage<N> {
+    type Buffer<T> = [MaybeUninit<T>; N];
 }
 
-impl<T> VecBuffer for [MaybeUninit<T>] {
-    type T = T;
-
-    fn as_vecview(vec: &VecInner<Self>) -> &VecView<Self::T> {
-        vec
-    }
-    fn as_mut_vecview(vec: &mut VecInner<Self>) -> &mut VecView<Self::T> {
-        vec
-    }
+/// Implementation of [`Storage`] for [`VecView`].
+pub enum VecViewStorage {}
+impl Storage for VecViewStorage {}
+impl SealedStorage for VecViewStorage {
+    type Buffer<T> = [MaybeUninit<T>];
```

Comment:
I think It would make sense to for these structs be moved in a common module for the crate, and rename the structs to `Owned` and `View`, so that the same structs can be used for other containers ?

I think it's clearer to use the same structs for all containers.

Reply:
good point, done!